### PR TITLE
feat(ir): add tile.ci / tensor.ci ops (arange) — #1095

### LINKED ISSUE #1095: [New Op] Add TCI (tile.ci / tensor.ci) — contiguous integer sequence generation
### Operation Level

Other (please specify): **both tile-level and tensor-level**. The op must be registered at both layers (`tile.ci` and `tensor.ci`) with a tensor-to-tile conversion, mirroring how other tensor ops lower to their tile counterparts.

### Proposed Name & Signature

IR-level C++ ops: `tile.ci` and `tensor.ci`.

Python IR / DSL wrappers — propose `ci(...)` (matching ISA) with an `arange(...)` convenience alias:

```python
# IR layer — python/pypto/ir/op/tile_ops.py, python/pypto/ir/op/tensor_ops.py
def ci(
    start: int | Expr,
    *,
    shape: Sequence[int | Expr] | None = None,      # required for tensor.ci; tile.ci may receive it via result_type
    dtype: DataType = DataType.INT32,
    descending: bool = False,
    span: Span | None = None,
) -> Call: ...

# DSL layer — python/pypto/language/op/tile_ops.py, python/pypto/language/op/tensor_ops.py
def ci(start: int | Scalar, shape: Sequence[IntLike], dtype: DataType = pl.INT32,
       descending: bool = False) -> Tile: ...         # tile.*
def ci(start: int | Scalar, shape: Sequence[IntLike], dtype: DataType = pl.INT32,
       descending: bool = False) -> Tensor: ...       # tensor.*

# Convenience alias
arange = ci   # so users can write pl.arange(0, shape=[4, 32])
```

### Semantics Description

`TCI` generates a contiguous integer sequence into a destination tile/tensor. For a linearized index `k` over the valid elements:

- Ascending: `dst[k] = start + k`
- Descending: `dst[k] = start - k`

The linearization order depends on tile layout (implementation-defined).

Reference: `pto-isa/docs/isa/TCI.md` and figure `pto-isa/docs/figures/isa/TCI.svg`.

**Inputs**

- `start`: scalar, same dtype as destination
- `shape`: output shape
- `dtype`: destination dtype
- `descending`: bool kwarg (default `false`)

**Output**

- Tile or tensor with shape `shape` and dtype `dtype`, populated per the formula above.

**Constraints (per ISA)**

- `dtype` must be one of `INT32`, `UINT32`, `INT16`, `UINT16`.
- Destination `Cols != 1` (column vectors not supported).
- Hardware uses `dst.GetValidCol()` as sequence length; does not consult `GetValidRow()`.
- Scalar `start` dtype must match destination dtype.

### Example Usage

```python
import pypto.language as pl

@pl.program
class Example:
    @pl.function(type=pl.FunctionType.InCore)
    def main_incore_0(self, out: pl.Out[pl.Tensor[[4, 32], pl.INT32]]) -> pl.Tensor[[4, 32], pl.INT32]:
        # Generate [0, 1, 2, ..., 31] per row (linearization order is layout-defined)
        idx: pl.Tensor[[4, 32], pl.INT32] = pl.tensor.ci(0, shape=[4, 32], dtype=pl.INT32)
        # Or equivalently via the arange alias
        idx2: pl.Tensor[[4, 32], pl.INT32] = pl.arange(0, shape=[4, 32], dtype=pl.INT32)
        return idx
```

A typical use is preparing the `idx` operand for `tile.sort32` / `tensor.sort32`, where the user currently has to materialize the 0..N-1 sequence manually.

### Hardware Support Considerations

- Maps directly to the existing PTO-AS instruction `pto.tci`. The backend codegen is already partially wired: `src/backend/common/pto_ops_common.cpp:456` (`MakeCiCodegenPTO`) and the registration at `src/backend/common/pto_ops_common.cpp:1566` target `tile.ci`, but the IR-level op is not registered (`ir.is_op_registered(\"tile.ci\") == False`). This issue is to add the missing IR/DSL/conversion layers so that codegen actually fires.
- No CPU fallback needed for this hardware-specific sequence generator.
- Runs in `MemorySpace::Vec` on Ascend 910B / A2A3 (aligned with other tile compute ops).

### Motivation / Use Case

1. **`sort32` usability.** `tile.sort32` / `tensor.sort32` require an `idx` tile/tensor initialized to a sequence (`[0, 1, ..., 31]` for FP16 src, `[0, 2, ..., 62]` for FP32 src). Without `arange`/`ci`, users must fabricate the sequence via orchestration, which is awkward and precludes a fully on-device kernel. See `tests/st/runtime/test_sort.py`, which currently constructs `idx` externally.
2. **Index generation for masking / gather.** Many kernels need a row/column index tile — `tensor.gather`, position-encoded ops, softmax length masks, etc. `arange` is a standard primitive in numpy/torch/triton/JAX; its absence forces users to fake it with external tensor inputs.
3. **Backend already supports it.** Codegen for `tile.ci → pto.tci` exists; the front-end and IR gap is the only thing blocking usage. Closing the loop is low-risk and unlocks immediate value.

### Additional Context

- ISA reference: `pto-isa/docs/isa/TCI.md` (ascending/descending contiguous integer sequence, `descending` attribute).
- Existing backend codegen: `src/backend/common/pto_ops_common.cpp:456` and `:1566`.
- Related tests that would benefit: `tests/st/runtime/test_sort.py` (`test_sort32_fp32`, `test_sort32_gather_fp32`).
- Framework analogues: `numpy.arange`, `torch.arange`, `tl.arange` (Triton), `jax.numpy.arange`.

**Suggested implementation scope (Phase A + Phase B of the `add-op` skill):**

- C++ `REGISTER_OP(\"tile.ci\")` with 1 arg (start scalar), `descending` bool attr, `Vec` output memory, dtype restricted to {INT16, UINT16, INT32, UINT32}.
- C++ `REGISTER_OP(\"tensor.ci\")` similarly (no memory spaces).
- Python IR wrappers in `python/pypto/ir/op/{tile,tensor}_ops.py`.
- Python DSL wrappers + `arange` alias in `python/pypto/language/op/{tile,tensor}_ops.py`.
- `RegisterSimple(\"tensor.ci\", \"tile.ci\")` in `src/ir/transforms/op_conversion_registry.cpp`.
- Unit tests + conversion tests.
- Docs update in `docs/en/dev/ir/05-operators.md` + `docs/zh-cn/...`.

### PR DESCRIPTION
**Pull request by Little-oil**
## Summary

Register TCI (contiguous integer sequence) at the IR / DSL / conversion layers so the already-wired `pto.tci` codegen actually fires. Before this PR, `ir.is_op_registered("tile.ci") == False` blocked all usage despite backend codegen being in place.

Fixes #1095

## Usage

```python
import pypto.language as pl

# Ascending: dst[k] = start + k
idx: pl.Tile[[1, 32], pl.INT32] = pl.tile.ci(0, [1, 32], dtype=pl.INT32)

# Descending: dst[k] = start - k
idx: pl.Tile[[1, 32], pl.INT32] = pl.tile.ci(31, [1, 32], dtype=pl.INT32, descending=True)

# Tensor variant (lowers to tile.ci via conversion pass)
idx: pl.Tensor[[1, 32], pl.INT32] = pl.tensor.ci(0, [1, 32], dtype=pl.INT32)

# arange alias
idx = pl.arange(0, [1, 32], dtype=pl.INT32)
```

Primary motivation: supply the `idx` operand for `tile.sort32` / `tensor.sort32` without fabricating it via orchestration.

## Inputs / Outputs

| Param         | Type                  | Notes                                                    |
| ------------- | --------------------- | -------------------------------------------------------- |
| `start`       | scalar                | same dtype as destination                                |
| `shape`       | `Sequence[IntLike]`   | innermost dim must be != 1                               |
| `dtype`       | `DataType`            | **`{INT16, INT32}` only** (see note below)               |
| `descending`  | `bool` (default `False`) | ascending vs. descending                              |

**Output**: tile/tensor of the given `shape` / `dtype` populated per the formula.

## Important ISA Constraint — only the first row is populated

Per `_deps/pto-isa/docs/isa/tile/ops/irregular-and-complex/tci.md` (Constraints section):

> The implementation uses `dst.GetValidCol()` as the sequence length and does not consult `dst.GetValidRow()`.

This means for shape `[M, N]` with `M > 1`, **only row 0 gets `[start, start±1, …, start±(N-1)]`**; rows 1..M-1 retain UB residue. Typical correct usage is `shape=[1, N]`, which matches sort32's `idx` operand. Multi-row arange is out of scope for this PR and should be built on top via broadcast (future work).

## Why dtype is narrowed to `{INT16, INT32}` (no UINT)

The ISA doc nominally allows `{INT16, UINT16, INT32, UINT32}` (`tci.md:96`), but the PTOAS lowering pipeline has a gap on the unsigned path:

1. MLIR `arith.constant` requires **signless** integer types (`i32`, not `ui32`).
2. Bridging signless → unsigned via `builtin.unrealized_conversion_cast` produces valid MLIR but PTOAS recovers the template parameter `T` from the SSA def chain, **ignoring the cast**.
3. The result: `TCI<Tile<..., uint32_t, ...>, int32_t, 0>(...)` — `T=int32_t` mismatches `TileData::DType=uint32_t`, triggering `static_assert<is_same<DType, T>>` failure in `_deps/pto-isa/include/pto/npu/a2a3/TCI.hpp:20`.
4. Official PTOAS ST (`_deps/pto-isa/tests/npu/a2a3/src/st/testcase/tci/tci_kernel.cpp`) only instantiates `int32_t` and `int16_t` — the unsigned path was never validated upstream.

Frontend UTs added (`test_tile_ci_rejects_uint_dtype`, `test_tensor_ci_rejects_uint_dtype`) guard this restriction. Re-enable UINT16/UINT32 once PTOAS supports it properly.

## Changes

- **C++ IR**: `REGISTER_OP("tile.ci")` in `src/ir/op/tile_ops/memory.cpp`; `REGISTER_OP("tensor.ci")` in `src/ir/op/tensor_ops/memory.cpp`.
- **Python IR**: `ci(...)` in `python/pypto/ir/op/{tile,tensor}_ops.py`.
- **Python DSL**: `ci(...)` + `arange` alias in `python/pypto/language/op/{tile,tensor}_ops.py`.
- **Conversion**: `RegisterSimple("tensor.ci", "tile.ci")` in `op_conversion_registry.cpp`.
- **Backend**: small `MakeCiCodegenPTO` tweak to emit DPS-form type annotations (`ins(%v {attr} : type) outs(%d : type)`).
- **Docs**: en/zh operator table entries.
- **Tests**: UTs for both ops, conversion test, ST (`tests/st/runtime/test_ci.py`).

## Test plan

- [x] Unit tests: `pytest tests/ut/ir/operators/ -v -k ci`
- [x] Conversion test: `pytest tests/ut/ir/transforms/ -v -k ci`
- [x] ST (PTOAS path on 910B): `pytest tests/st/runtime/test_ci.py` — ascending start=0, ascending start=10, descending, all passed
- [x] `pre-commit run --all-files` passed

### CODE REVIEW

File: src/ir/op/tensor_ops/memory.cpp [hunk 1]
```diff
@@ -393,6 +393,79 @@ REGISTER_OP("tensor.full")
...
+  CHECK(!shape.empty()) << "tensor.ci requires non-empty shape";
+
+  // ISA constraint: innermost dim Cols != 1.
+  if (auto last_const = As<ConstInt>(shape.back())) {
+    CHECK(last_const->value_ != 1) << "tensor.ci requires the innermost dimension (Cols) to be != 1, got "
+                                   << last_const->value_;
+  }
+
+  (void)kwargs;  // descending is optional bool kwarg, no validation needed beyond type.
+  return std::make_shared<TensorType>(shape, dtype);
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/tensor_ops/memory.cpp:484]: Fixed in `a8feeb41` — tensor.ci enforces leading dims == 1 for compile-time constants.

File: tests/ut/ir/operators/test_tile_ops.py [hunk 2]
```diff
@@ -2543,5 +2543,42 @@ def test_tile_add_symbolic_shape_mismatch_shows_var_names(self):
...
+            tile.ci(0, [2, 16], dtype=DataType.UINT32)
+
+    def test_tile_ci_rejects_cols_equal_one(self):
+        with pytest.raises(ValueError, match="innermost dimension"):
+            tile.ci(0, [32, 1], dtype=DataType.INT32)
+
+    def test_tile_ci_rejects_start_dtype_mismatch(self):
+        span = ir.Span.unknown()
+        start = ir.Var("s", ir.ScalarType(DataType.INT16), span)
+        with pytest.raises(ValueError, match="start.*dtype"):
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply tests/ut/ir/operators/test_tile_ops.py]: Fixed in `a8feeb41` — pytest.raises(match=...) now uses raw strings (RUF043).

File: src/ir/op/tile_ops/memory.cpp [hunk 3]
```diff
@@ -373,6 +373,62 @@ TypePtr DeduceTileFullType(const std::vector<ExprPtr>& args,
...
+  CHECK(last_dim && last_dim->value_ != 1)
+      << "The operator " << op_name << " requires the innermost dimension (Cols) to be != 1, got "
+      << (last_dim ? last_dim->value_ : -1);
+
+  // descending kwarg is optional and defaults to false; just validate type if provided.
+  (void)GetKwarg<bool>(kwargs, "descending");
+
+  TileView tile_view;
+  tile_view.valid_shape = tile_shape;
+  return std::make_shared<TileType>(tile_shape, dtype, std::nullopt, tile_view);
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/tile_ops/memory.cpp:455]: Fixed in `a8feeb41` — tile.ci now enforces leading dims == 1 and rejects multi-row shapes.

File: python/pypto/ir/op/tile_ops.py [hunk 4]
```diff
@@ -479,6 +479,45 @@ def full(
...
+    """Generate a contiguous integer sequence into a tile (pto.tci).
+
+    For a linearized index ``k`` over the valid elements:
+    - Ascending: ``dst[k] = start + k``
+    - Descending: ``dst[k] = start - k``
+
+    Args:
+        start: Starting integer (plain int or a scalar Expr). Its dtype must match ``dtype``.
+        shape: Destination tile shape (static).
+        dtype: Destination dtype. Must be one of {INT16, INT32}.
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply python/pypto/ir/op/tile_ops.py]: Fixed in `a8feeb41` — docstrings clarify that pto.tci only populates the first row and leading dims must be 1.

File: src/ir/op/tensor_ops/memory.cpp [hunk 5]
```diff
@@ -393,6 +393,79 @@ REGISTER_OP("tensor.full")
...
+      shape.emplace_back(std::make_shared<TupleGetItemExpr>(args[1], static_cast<int>(i), args[1]->span_));
+    }
+  }
+  CHECK(!shape.empty()) << "tensor.ci requires non-empty shape";
+
+  // ISA constraint: innermost dim Cols != 1.
+  if (auto last_const = As<ConstInt>(shape.back())) {
+    CHECK(last_const->value_ != 1) << "tensor.ci requires the innermost dimension (Cols) to be != 1, got "
+                                   << last_const->value_;
+  }
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/tensor_ops/memory.cpp:471]: Fixed in `a8feeb41` — tensor.ci now rejects multi-row compile-time shapes in DeduceTensorCiType.

File: src/ir/op/tile_ops/memory.cpp [hunk 6]
```diff
@@ -373,6 +373,62 @@ TypePtr DeduceTileFullType(const std::vector<ExprPtr>& args,
...
+  CHECK(!tile_shape.empty()) << "The operator " << op_name << " requires non-empty shape";
+
+  // ISA constraint: destination Cols != 1 (column vectors not supported by pto.tci).
+  auto last_dim = As<ConstInt>(tile_shape.back());
+  CHECK(last_dim && last_dim->value_ != 1)
+      << "The operator " << op_name << " requires the innermost dimension (Cols) to be != 1, got "
+      << (last_dim ? last_dim->value_ : -1);
+
+  // descending kwarg is optional and defaults to false; just validate type if provided.
+  (void)GetKwarg<bool>(kwargs, "descending");
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/tile_ops/memory.cpp]: Fixed in `a8feeb41` — GetKwarg<bool>("descending", false) now has a default so the kwarg is optional.

File: src/ir/op/tile_ops/memory.cpp [hunk 7]
```diff
@@ -373,6 +373,62 @@ TypePtr DeduceTileFullType(const std::vector<ExprPtr>& args,
...
+                                 << " must be positive, got " << const_int->value_;
+    tile_shape.push_back(make_tuple->elements_[i]);
+  }
+  CHECK(!tile_shape.empty()) << "The operator " << op_name << " requires non-empty shape";
+
+  // ISA constraint: destination Cols != 1 (column vectors not supported by pto.tci).
+  auto last_dim = As<ConstInt>(tile_shape.back());
+  CHECK(last_dim && last_dim->value_ != 1)
+      << "The operator " << op_name << " requires the innermost dimension (Cols) to be != 1, got "
+      << (last_dim ? last_dim->value_ : -1);
```
[the remark this answers is not in the document]
Comment (Little-oil) [reply src/ir/op/tile_ops/memory.cpp:438]: Fixed in `a8feeb41` — tile.ci now rejects multi-row shapes (all leading dims must be 1).

**Comment by Little-oil** [comment]
PTOASv0.29存在cube核内循环bug+pto.mrgsort接口bug，等bug解决后删除test.sort的skipped可以合入

**Review by lyfne123 — APPROVED** [review]
